### DISCUSSION TITLE
How to install programmatically in a container?

### DISCUSSION THREAD

**Discussion by miunsag** [microsoft/vscode-cpptools#11154] [Q&A]
Hello all,

I am trying to build a sort of devcontainer that includes this extension, but I have not been able to find any clear programmatic way to include this extension or the pack.

I am expecting something like 

`code-server --install-extension ms-vscode.cpptools-extension-pack`

in the Dockerfile, for example:

```Dockerfile
ARG __base_img=ubuntu
FROM ${__base_img}

RUN apt-get update &&\
    apt-get install --no-install-recommends -qy \
        bash \
        build-essential \
        ca-certificates \
        cmake \
        curl \
        gdb \
        libarchive-tools \
        libssl-dev \
        net-tools \
        &&\
    apt-get clean &&\
    rm -rf /var/lib/apt/lists/* &&\
    curl -fsSL https://code-server.dev/install.sh | sh

RUN groupadd vscodegrp &&\
    useradd -m -s /bin/bash -g vscodegrp vscode

USER vscode
RUN code-server --install-extension ms-vscode.cpptools-extension-pack

ENTRYPOINT [ "code-server" ]
```

Now, if I am looking at extensions like ms-python.python I see they come down with ease

```sh
vscode@bb633b4e60c1:~/.vscode-server/extensions$ code-server --install-extension ms-python.python --force
Installing extensions...
Installing extension 'ms-python.python'...
```

while this one doesn't:

```sh
vscode@bb633b4e60c1:~/.vscode-server/extensions$ code-server --install-extension cpptools-extension-pack --force
Installing extensions...
Extension 'cpptools-extension-pack' not found.
Make sure you use the full extension ID, including the publisher, e.g.: ms-dotnettools.csharp
Failed Installing Extensions: cpptools-extension-pack
```

I have found some alternative instructions telling me to download a .vsix file, but I am only able to download it manually, I have no idea how to do it programmatically in a build pipeline for the docker image.

How can I install programmatically, i.e. 100% unattended, the latest version in a Dockerfile?

Thanks

**Comment by Colengms** [comment]
Hi @miunsag .  The extension ID of the extension pack is `ms-vscode.cpptools-extension-pack`.  You can find this by finding the extension pack (or any extension) in the list of extensions on the marketplace tab, selecting it, selecting the gear icon, and selecting `Copy Extension ID`.

![image](https://github.com/microsoft/vscode-cpptools/assets/49173979/9ce04d53-aca9-4af4-b308-a92c83176398)

That would put it into your clipboard.  Pasting, I see that it's:

`ms-vscode.cpptools-extension-pack`

**Reply by miunsag** [reply]
Sorry, but that was exactly what I was doing, but the commandline result is negative.

Try this out and you will see that the build fails at the mentioned point.

Current Dockerfile:

```Dockerfile
ARG __base_img=ubuntu
FROM ${__base_img}
ARG __project_dir=/pj

RUN apt-get update &&\
    apt-get install --no-install-recommends -qy \
        bash\
        build-essential\
        ca-certificates\
        clang-format\
        clang-tidy\
        clang-tools\
        clang\
        clangd\
        cmake\
        cppcheck\
        curl\
        flawfinder\
        gdb\
        git\
        libarchive-tools\
        libc++-dev\
        libc++1\
        libc++abi-dev\
        libc++abi1\
        libclang-dev\
        libclang1\
        liblldb-dev\
        libllvm-ocaml-dev\
        libomp-dev\
        libomp5\
        libssl-dev\
        lld\
        lldb\
        llvm-dev\
        llvm-runtime\
        llvm\
        net-tools\
        python3-clang\
        python3-pip\
        &&\
    pip install --no-cache-dir --upgrade pip && \
    pip install --no-cache-dir lizard &&\
    apt-get clean &&\
    rm -rf /var/lib/apt/lists/* &&\
    cd /dev/shm &&\
    curl -fsSL https://code-server.dev/install.sh | sh && \
    groupadd vscodegrp &&\
    useradd -m -s /bin/bash -g vscodegrp vscode &&\
    mkdir -p ${__project_dir} &&\
    chown vscode:vscodegrp ${__project_dir}

USER vscode

# TODO: how to install cpp tools extension pack, as this does not work?
# Opened discussion: https://github.com/microsoft/vscode-cpptools/discussions/11154
# RUN code-server --install-extension ms-vscode.cpptools-extension-pack
RUN code-server --install-extension sonarsource.sonarlint-vscode &&\
    code-server --install-extension xaver.clang-format &&\
    code-server --install-extension formulahendry.code-runner &&\
    code-server --install-extension oderwat.indent-rainbow &&\
    code-server --install-extension ban.spellright &&\
    code-server --install-extension hediet.vscode-drawio &&\
    code-server --install-extension brobeson.vscode-lizard

EXPOSE 8080

VOLUME ${__project_dir}

ENTRYPOINT [ "code-server" ]
```

**Reply by sean-mcmanus** [reply]
What is the error? Does installing ms-vscode.cpptools work?

**Reply by miunsag** [reply]
Pasting below the shell experience.

```sh
code-server --install-extension ms-vscode.cpptools-extension-pack
```

```text
Installing extensions...
Extension 'ms-vscode.cpptools-extension-pack' not found.
Make sure you use the full extension ID, including the publisher, e.g.: ms-dotnettools.csharp
Failed Installing Extensions: ms-vscode.cpptools-extension-pack
vscode@4c72b6709670:/dev/shm$ echo $?
0
```

Same for cpptools

```text
vscode@4c72b6709670:/dev/shm$ code-server --install-extension ms-vscode.cpptools
Installing extensions...
Extension 'ms-vscode.cpptools' not found.
Make sure you use the full extension ID, including the publisher, e.g.: ms-dotnettools.csharp
Failed Installing Extensions: ms-vscode.cpptools
```

However, if I am not concatenating these with && and leave them fail, then, when VSCode attaches to the container it installs again and this time it seems the extensions are up.

This may suffice for the purpose, however it seems that the CLI installation at container build time has some issues and maybe that is why VSCode refreshes again everything at start time (?) I wonder what would happen if I spun the container without internet connectivity.

**Reply by Colengms** [reply]
Hi @miunsag .  You might consider reporting this in [the vscode repo](https://github.com/microsoft/vscode/issues) .  Extension (and pack) installation is outside of the control of the C/C++ Extension itself.  If there is an issue here, it would likely need to be addressed in VS Code.  It's possible that installation of extension packs is not supported via command line.  Someone in the VS Code repo would know for sure.
